src/wallet.rs

```diff
@@ -0,0 +1,338 @@
+				panic!("Tried to spend a non-witness funding output. This must not ever happen.");
+			}
+		}
+
+		if !locked_wallet.sign(&mut psbt, SignOptions::default())? {
+			return Err(Error::FundingTxCreationFailed);
+		}
+
+		Ok(psbt.extract_tx())
+	}
+
+	pub(crate) fn get_new_address(&self) -> Result<bitcoin::Address, Error> {
+		let address_info = self.inner.lock().unwrap().get_internal_address(AddressIndex::New)?;
+		Ok(address_info.address)
+	}
+
+	#[cfg(any(test))]
+	pub(crate) fn get_balance(&self) -> Result<bdk::Balance, Error> {
+		Ok(self.inner.lock().unwrap().get_balance()?)
+	}
+
+	fn estimate_fee_rate(&self, confirmation_target: ConfirmationTarget) -> FeeRate {
+		let mut locked_fee_rate_cache = self.fee_rate_cache.lock().unwrap();
+		let num_blocks = num_blocks_from_conf_target(confirmation_target);
+
+		// We'll fall back on this, if we really don't have any other information.
+		let fallback_rate = fallback_fee_from_conf_target(confirmation_target);
+
+		let locked_runtime = self.tokio_runtime.read().unwrap();
+		if locked_runtime.as_ref().is_none() {
+			log_error!(self.logger, "Failed to update fee rate estimation: No runtime.");
+			return *locked_fee_rate_cache.get(&confirmation_target).unwrap_or(&fallback_rate);
+		}
+
+		let est_fee_rate = tokio::task::block_in_place(move || {
+			locked_runtime
+				.as_ref()
+				.unwrap()
+				.handle()
+				.block_on(async move { self.blockchain.estimate_fee(num_blocks).await })
+		});
+
+		match est_fee_rate {
+			Ok(rate) => {
+				locked_fee_rate_cache.insert(confirmation_target, rate);
+				log_trace!(
+					self.logger,
+					"Fee rate estimation updated: {} sats/kwu",
+					rate.fee_wu(1000)
+				);
+				rate
+			}
+			Err(e) => {
+				log_error!(self.logger, "Failed to update fee rate estimation: {}", e);
+				*locked_fee_rate_cache.get(&confirmation_target).unwrap_or(&fallback_rate)
+			}
+		}
+	}
+}
+
+impl<D> FeeEstimator for Wallet<D>
+where
+	D: BatchDatabase,
+{
+	fn get_est_sat_per_1000_weight(&self, confirmation_target: ConfirmationTarget) -> u32 {
+		(self.estimate_fee_rate(confirmation_target).fee_wu(1000) as u32)
+			.max(FEERATE_FLOOR_SATS_PER_KW)
+	}
+}
+
+impl<D> BroadcasterInterface for Wallet<D>
+where
+	D: BatchDatabase,
+{
+	fn broadcast_transaction(&self, tx: &Transaction) {
+		let locked_runtime = self.tokio_runtime.read().unwrap();
+		if locked_runtime.as_ref().is_none() {
+			log_error!(self.logger, "Failed to broadcast transaction: No runtime.");
+			panic!("Failed to broadcast transaction: No runtime.");
+		}
+
+		let res = tokio::task::block_in_place(move || {
+			locked_runtime
+				.as_ref()
+				.unwrap()
+				.block_on(async move { self.blockchain.broadcast(tx).await })
+		});
+
+		match res {
+			Ok(_) => {}
+			Err(err) => {
+				log_error!(self.logger, "Failed to broadcast transaction: {}", err);
+				panic!("Failed to broadcast transaction: {}", err);
```

Comment:
Really not sure we should panic here. Currently running into panics when trying to force close a channel due to the server not accepting a duplicate without a fee raise. Funnily enough, we'll get the error arising in `bitcoind` right handed back down the chain and receive it as a non-descript HTTP 400 error:

```
Failed to broadcast transaction: Esplora(Reqwest(reqwest::Error { kind: Status(400), url: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(0.0.0.0)), port: Some(50292), path: "/tx", query: None, fragment: None } }))
```

Corresponding line in the `electrsd` log is:
```
WARN - HttpError(400, "sendrawtransaction RPC error: {\"code\":-26,\"message\":\"insufficient fee, rejecting replacement a284e16509213767025b961995883788465102a7a691f3736cc6ca8ba626d090; new feerate 0.00001234 BTC/kvB <= old feerate 0.00001343 BTC/kvB\"}")
```